code/espurna/ac_switch.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+
+AC FREQUENCY DETECTION SWITCH MODULE
+
+*/
+
+#include "ac_switch.h"
+#include "espurna.h"
+
+namespace espurna {
+
+AcSwitch::AcSwitch(uint8_t pin, unsigned long timeout_ms, unsigned long min_pulses, int frequency) :
+    _pin(pin),
+    _timeout_ms(timeout_ms),
+    _min_pulses(min_pulses)
+{
+    if (frequency == 50) {
+        // 10000us period for 50Hz with CHANGE interrupt. Allow 3.5% tolerance.
+        _min_filter_period = 9650;
+        _max_filter_period = 10350;
+    } else if (frequency == 60) {
+        // ~8333us period for 60Hz. Allow 3.5% tolerance.
+        _min_filter_period = 8041;
+        _max_filter_period = 8625;
+    }
+
+    // INPUT_PULLUP is usually safer if the optocoupler pulls to ground.
+    // If the schematic is different, INPUT might be required.
+    pinMode(_pin, INPUT_PULLUP);
+    attachInterrupt(_pin, [this]() { this->_isr(); }, CHANGE);
```

Comment:
iirc there is caching issue when using lambda, eventually it'd prune it out of the IRAM and cause exception when sdk isr handler attempts calling it